packages/snaps-controllers/src/snaps/location/npm.ts

```diff
@@ -0,0 +1,353 @@
+import {
+  assertIsSemVerVersion,
+  assertIsSnapManifest,
+  deepClone,
+  DEFAULT_REQUESTED_SNAP_VERSION,
+  getTargetVersion,
+  isValidUrl,
+  NpmSnapIdStruct,
+  SemVerRange,
+  SemVerVersion,
+  SnapManifest,
+  VFile,
+} from '@metamask/snaps-utils';
+import { assert, assertStruct, isObject } from '@metamask/utils';
+import concat from 'concat-stream';
+import createGunzipStream from 'gunzip-maybe';
+import pump from 'pump';
+import { ReadableWebToNodeStream } from 'readable-web-to-node-stream';
+import { Readable, Writable } from 'stream';
+import { extract as tarExtract } from 'tar-stream';
+
+import { SnapLocation } from './location';
+
+const DEFAULT_NPM_REGISTRY = 'https://registry.npmjs.org';
+
+interface NpmMeta {
+  registry: URL;
+  packageName: string;
+  requestedRange: SemVerRange;
+  version?: string;
+  fetch: typeof fetch;
+}
+export interface NpmOptions {
+  /**
+   * @default DEFAULT_REQUESTED_SNAP_VERSION
+   */
+  versionRange?: SemVerRange;
+  /**
+   * The function used to fetch data.
+   *
+   * @default globalThis.fetch
+   */
+  fetch?: typeof fetch;
+  /**
+   * Whether to allow custom NPM registries outside of {@link DEFAULT_NPM_REGISTRY}.
+   *
+   * @default false
+   */
+  allowCustomRegistries?: boolean;
+}
+
+export default class NpmLocation implements SnapLocation {
+  private readonly meta: NpmMeta;
+
+  private validatedManifest?: VFile<SnapManifest>;
+
+  private files?: VFile[];
+
+  constructor(url: URL, opts: NpmOptions = {}) {
+    const allowCustomRegistries = opts.allowCustomRegistries ?? false;
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    const fetch = opts.fetch ?? globalThis.fetch;
+    const requestedRange = opts.versionRange ?? DEFAULT_REQUESTED_SNAP_VERSION;
+
+    assertStruct(url.toString(), NpmSnapIdStruct, 'Invalid Snap Id: ');
+
+    let registry: string | URL;
+    if (
+      url.host === '' &&
+      url.port === '' &&
+      url.username === '' &&
+      url.password === ''
+    ) {
+      registry = new URL(DEFAULT_NPM_REGISTRY);
+    } else {
+      registry = 'https://';
+      if (url.username) {
+        registry += url.username;
+        if (url.password) {
+          registry += `:${url.password}`;
+        }
+        registry += '@';
+      }
+      registry += url.host;
+      registry = new URL(registry);
+      assert(
+        allowCustomRegistries,
+        new TypeError(
```

Comment:
Not sure about the use of `TypeError` here.

Reply:
This error might possibly propagate to the DApp doing `wallet_enable` with a custom registry, and it's expected to crop up so not really `AssertionError` either.

If you have a better error, I'm open.